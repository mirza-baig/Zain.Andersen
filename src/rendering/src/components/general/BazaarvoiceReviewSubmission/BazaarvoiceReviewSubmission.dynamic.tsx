// Global
import { useSitecoreContext, withDatasourceCheck } from '@sitecore-jss/sitecore-jss-nextjs';
import { ThemeFile, ThemeName, useTheme } from 'lib/context/ThemeContext';
import { useBVScript } from 'lib/utils/use-bv-script';
import { useRouter } from 'next/router';
import { useEffect, useState } from 'react';
import { Feature } from 'src/.generated/Feature.EnterpriseWeb.model';

// Components
import { Component } from 'src/helpers/Component';
import ImageWrapper from 'src/helpers/Media/ImageWrapper';
import { BazaarvoiceReviewSubmissionTheme } from './BazaarvoiceReviewSubmission.theme';

import { getHeadingLevel } from 'lib/utils';
import { Headline } from 'src/helpers/Headline';
import { RichTextWrapper } from 'src/helpers/RichTextWrapper';
import { ProductByBVIdQueryResult } from 'src/pages/api/aw/bazaarvoice-product-by-bvid';
import { Item } from '@sitecore-jss/sitecore-jss-nextjs';

export type BazaarvoiceReviewSubmissionProps =
  Feature.EnterpriseWeb.AndersenWindows.Components.General.BazaarvoiceReviewSubmission.BazaarvoiceReviewSubmission;

const RenderReviewButton = (
  props: ProductByBVIdQueryResult,
  themeData: ThemeFile[ThemeName] | undefined
) => {
  if (props === null) {
    return <></>;
  }

  const imageField = {
    value: props?.productImage,
  };

  return (
    <>
      <div className={themeData.classes.bazaarvoiceImageWrapper}>
        <ImageWrapper image={imageField}></ImageWrapper>
      </div>
      <div className={themeData.classes.bazaarvoiceProductContent}>
        <h3 className={themeData.classes.bazaarvoiceProductHeadline}>{props.productName?.value}</h3>
        <button
          onClick={() => SubmitReviewClick(props.bazaarvoiceProductId?.value)}
          className={themeData.classes.buttonClass}
          title="Start Rating"
        >
          Start Rating
        </button>
      </div>
    </>
  );
};

const SubmitReviewClick = (externalId: string) => {
  // @ts-ignore $BV will be available after the page loads.
  $BV.ui('rr', 'submit_review', { productId: externalId });
};

const RenderAllReviewButtons = (
  props: ProductByBVIdQueryResult[],
  themeData: ThemeFile[ThemeName] | undefined
) => {
  if (props === null || props?.length === 0 || !Array.isArray(props)) {
    return <></>;
  }

  return (
    <div className={themeData.classes.bazaarvoiceContainerWrapper}>
      {props.map((_item, _index) => (
        <div key={_index} className={themeData.classes.bazaarvoiceProductWrapper}>
          {RenderReviewButton(_item, themeData)}
        </div>
      ))}
    </div>
  );
};

const BazaarvoiceReviewSubmission = (props: BazaarvoiceReviewSubmissionProps) => {
  const { themeData } = useTheme(BazaarvoiceReviewSubmissionTheme);
  const router = useRouter();
  const sitecoreContext = useSitecoreContext();
  const language = sitecoreContext.sitecoreContext.language;

  const [searchResults, setSearchResults] = useState<ProductByBVIdQueryResult[]>([]);

  // Add the bazaarvoice script
  useBVScript();

  useEffect(() => {
    async function getProductsByBVProductId() {
      const urlSearchParams = new URLSearchParams(router.asPath.split('?')[1]);
      const productIds = urlSearchParams.get('externalIds');
      const sourceIds = props?.fields?.bazaarvoiceProducts?.map((product: Item) => product.id);

      if (sourceIds?.length > 0 && productIds) {
        try {
          const response = await fetch('/api/aw/bazaarvoice-product-by-bvid/', {
            method: 'POST',
            body: JSON.stringify({
              sourceIds: sourceIds,
              productIds: productIds.split(','),
              language: language,
            }),
            headers: {
              'Content-Type': 'application/json',
            },
          });

          if (response.ok) {
            const data = await response.json();
            setSearchResults(data.products);
          } else {
            console.error('Error:', response.statusText);
          }
        } catch (error) {
          console.error('Error occurred:', error);
        }
      }
    }
    getProductsByBVProductId();
  }, [router.asPath, router.isReady]);

  return (
    <Component variant="lg" dataComponent="general/bazaarvoicereviewsubmission" {...props}>
      <div className={themeData.classes.componentWrapperClass}>
        <Headline
          useTag={getHeadingLevel('h2', props.fields?.headlineLevel)}
          classes={themeData.classes.headlineContainer}
          {...props}
        />
        <RichTextWrapper
          field={props?.fields?.thankYouMessaging}
          classes={themeData.classes.thankYouMessaging}
        />
        {RenderAllReviewButtons(searchResults, themeData)}
      </div>
    </Component>
  );
};

export default withDatasourceCheck()<BazaarvoiceReviewSubmissionProps>(BazaarvoiceReviewSubmission);
