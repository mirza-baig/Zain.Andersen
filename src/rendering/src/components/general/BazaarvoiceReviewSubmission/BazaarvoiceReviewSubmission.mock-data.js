const defaultData = {
  rendering: {
    dataSource: 'sampledatasource',
    componentName: 'BazaarvoiceReviewSubmission',
  },
  fields: {
    "thankYouMessaging": {
      "value": "body copy of bv review submission"
    },
    "bazaarvoiceProducts": [
      {
        "id": "a971c447-5837-447c-801a-e9b970bf3d73",
        "url": "https://cm.local.ew.andersencorp.com/AndersenCorporation/AndersenWindows/Global/Products/Products/Big-Doors",
        "name": "Big Doors",
        "displayName": "Big Doors",
        "fields": {},
        "templateId": "15665680-e4a6-45ba-98e8-4b46a33b7a2d",
        "templateName": "Product Folder"
      },
      {
        "id": "1071e4d3-9dc2-4285-895b-8dccd34f8bc7",
        "url": "https://cm.local.ew.andersencorp.com/AndersenCorporation/AndersenWindows/Global/Products/Products/Casement-Windows",
        "name": "Casement Windows",
        "displayName": "Casement Windows",
        "fields": {},
        "templateId": "15665680-e4a6-45ba-98e8-4b46a33b7a2d",
        "templateName": "Product Folder"
      },
      {
        "id": "aaac67a1-e0d9-4379-a884-57af3b75e166",
        "url": "https://cm.local.ew.andersencorp.com/AndersenCorporation/AndersenWindows/Global/Products/Products/Windows",
        "name": "Windows",
        "displayName": "Windows",
        "fields": {},
        "templateId": "15665680-e4a6-45ba-98e8-4b46a33b7a2d",
        "templateName": "Product Folder"
      }
    ],
    "sectionId": {
      "value": ""
    },
    "componentSpacing": null,
    "headlineText": {
      "value": "headline!"
    },
    "eventName": {
      "value": ""
    },
    "eventType": {
      "value": ""
    },
    "eventZone": {
      "value": ""
    }
  },
};

export default defaultData;
