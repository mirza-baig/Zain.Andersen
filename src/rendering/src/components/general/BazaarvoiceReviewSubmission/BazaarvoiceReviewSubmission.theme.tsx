// Lib
import { ThemeFile } from 'lib/context/ThemeContext';
import { ButtonPrimaryClasses } from 'src/helpers/Button/buttons/btn--primary';

export const BazaarvoiceReviewSubmissionTheme: ThemeFile = {
  aw: {
    classes: {
      componentWrapperClass: 'col-span-12 lg:grid lg:grid-cols-12',
      headlineContainer: 'col-span-12 text-s md:text-m font-heavy text-theme-text pb-xs',
      thankYouMessaging:
        'col-span-12 text-body text-theme-text font-medium text-theme-body text-body',
      bazaarvoiceContainerWrapper: 'col-start-2 col-span-10 py-m lg:px-m',
      bazaarvoiceImageWrapper: 'w-[100px] lg:inline-block mx-auto lg:mx-0 pr-s mb-s lg:mb-0',
      bazaarvoiceProductContent:
        'inline-block flex flex-col items-center lg:items-start content-center align-middle place-content-center',
      bazaarvoiceProductHeadline: 'text-theme-text text-s mb-xs',
      bazaarvoiceProductWrapper: 'border border-[#b9b9b9] p-s mb-xs lg:flex lg:content-center',
      buttonClass: ButtonPrimaryClasses('aw'),
    },
  },
  rba: {
    classes: {
      componentWrapperClass: 'col-span-12 lg:grid lg:grid-cols-12',
      headlineContainer: 'col-span-12 text-s md:text-m font-heavy text-theme-text pb-xs',
      thankYouMessaging:
        'col-span-12 text-body text-theme-text font-medium text-theme-body text-body',
      bazaarvoiceContainerWrapper: 'col-start-2 col-span-10 py-m lg:px-m',
      bazaarvoiceImageWrapper: 'w-[100px] lg:inline-block mx-auto lg:mx-0 pr-s mb-s lg:mb-0',
      bazaarvoiceProductContent:
        'inline-block flex flex-col items-center lg:items-start content-center align-middle place-content-center',
      bazaarvoiceProductHeadline: 'text-theme-text text-s mb-xs',
      bazaarvoiceProductWrapper: 'border border-[#b9b9b9] p-s mb-xs lg:flex lg:content-center',
      buttonClass: ButtonPrimaryClasses('rba'),
    },
  },
};
