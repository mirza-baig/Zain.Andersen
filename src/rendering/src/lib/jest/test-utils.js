// Global
import { render, fireEvent, waitFor } from '@testing-library/react';
import { ComponentPropsContext } from '@sitecore-jss/sitecore-jss-nextjs';
import { FastSitecoreContext } from 'lib/overrides/FastSitecoreContext';

// Lib
import { createSampleComponentFactory } from 'lib/mocks/mock-placeholder';

export const eeSitecoreContext = {
  route: {
    name: 'Foo',
    placeholders: {},
  },
  context: {
    pageEditing: true,
    site: {
      name: 'JssTestWeb',
    },
    language: 'en',
  },
};

export const renderComponent = (
  Component,
  {
    componentProps = {},
    staticProps = {},
    componentFactory = createSampleComponentFactory(),
    sitecoreContext = {
      route: {
        name: 'Foo',
        placeholders: {},
      },
    },
  }
) => {
  const mockLayoutData = {
    sitecore: sitecoreContext,
  };

  return render(
    <ComponentPropsContext value={staticProps}>
      <FastSitecoreContext componentFactory={componentFactory} layoutData={mockLayoutData}>
        <Component {...componentProps} />
      </FastSitecoreContext>
    </ComponentPropsContext>
  );
};

export const snapshot = (Component, mockData) => {
  const component = renderComponent(Component, mockData);
  expect(component.container.firstChild).toMatchSnapshot();
  return component;
};

export const hasDataComponent = async (component, name) => {
  await waitFor(() => {
    const nodes = component.baseElement.querySelectorAll(`[data-component="${name}"]`);
    return expect(nodes.length).toBe(1);
  });
};

export const mockDateAs = (epochMs) => {
  if (typeof epochMs !== 'number') {
    console.error(
      'mockDateAs only accepts epoch values in milliseconds to prevent timezones from factoring in to Date creation.'
    );
  }

  jest.spyOn(global.Date, 'now').mockImplementation(() => new Date(epochMs).valueOf());
};

export const testFieldError = async (Component, mockData, { label, value, error }) => {
  const component = renderComponent(Component, mockData);
  const field = component.getByLabelText(label);
  fireEvent.change(field, { target: { value } });
  field.blur();
  await waitFor(() => {
    const err = component.queryByText(error);
    return expect(err).not.toBe(null);
  });
};
