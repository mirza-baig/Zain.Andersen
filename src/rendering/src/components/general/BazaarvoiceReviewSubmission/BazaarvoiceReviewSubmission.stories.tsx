// Global
import { Meta, Story } from '@storybook/react';
import { graphql } from 'msw';

// Local
import BazaarvoiceReviewSubmission, {
  BazaarvoiceReviewSubmissionProps,
} from './BazaarvoiceReviewSubmission.dynamic';
import defaultData from './BazaarvoiceReviewSubmission.mock-data';

export default {
  title: 'General/BazaarvoiceReviewSubmission',
  component: BazaarvoiceReviewSubmission,
} as Meta;
const Template: Story<BazaarvoiceReviewSubmissionProps> = (props) => (
  <BazaarvoiceReviewSubmission {...props} />
);
export const Default = Template.bind({});
Default.args = defaultData;
Default.parameters = {
  nextRouter: {
    path: '',
    asPath: '/?externalIds=100-CAS,400-CAS',
    query: {},
  },
  msw: {
    handlers: [
      // @ts-ignore - default param req used in definition.
      graphql.query('ProductByBVIdQuery', (req, res, ctx) => {
        return res(
          ctx.data({
            search: {
              total: 2,
              pageInfo: {
                endCursor: 'Mg==',
                hasNext: false,
              },
              results: [
                {
                  bazaarvoiceProductId: {
                    value: '100-CAS',
                  },
                  productName: {
                    value: '100 SERIES CASEMENT WINDOW',
                  },
                  productImage: {
                    src: 'https://place-puppy.com/200x200',
                    height: 200,
                    width: 200,
                    alt: '200x200 puppy',
                  },
                },
                {
                  bazaarvoiceProductId: {
                    value: '400-CAS',
                  },
                  productName: {
                    value: '400 SERIES FRENCHWOOD GLIDING PATIO DOOR',
                  },
                  productImage: {
                    src: 'https://place-puppy.com/100x100',
                    height: 100,
                    width: 100,
                    alt: '100x100 puppy',
                  },
                },
              ],
            },
          })
        );
      }),
    ],
  },
};
