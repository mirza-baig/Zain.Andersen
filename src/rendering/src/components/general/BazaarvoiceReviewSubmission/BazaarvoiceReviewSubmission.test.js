// Global
import { hasDataComponent, snapshot } from 'lib/jest/test-utils';
// Local
import BazaarvoiceReviewSubmission from './BazaarvoiceReviewSubmission.dynamic';
import defaultData from './BazaarvoiceReviewSubmission.mock-data';

jest.mock('next/router', () => ({
  useRouter() {
    return {
      route: '/',
      pathname: '',
      query: '',
      asPath: '',
      push: jest.fn(),
      events: {
        on: jest.fn(),
        off: jest.fn(),
      },
      beforePopState: jest.fn(() => null),
      prefetch: jest.fn(() => null),
    };
  },
}));

it('renders correctly', async () => {
  const component = snapshot(BazaarvoiceReviewSubmission, { componentProps: defaultData });
  await hasDataComponent(component, 'general/bazaarvoicereviewsubmission');
});
