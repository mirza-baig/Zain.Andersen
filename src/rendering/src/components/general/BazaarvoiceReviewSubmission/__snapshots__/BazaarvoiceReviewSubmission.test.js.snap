// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  data-component="general/bazaarvoicereviewsubmission"
  id="id1695242607"
>
  <div
    class="my-8 lg:mx-auto md:max-w-screen-lg theme-white"
  >
    <div
      class="section-grid grid grid-cols-2 md:grid-cols-12 px-m gap-s md:gap-s "
    >
      <div
        class="col-span-12 lg:grid lg:grid-cols-12"
      >
        <div
          class="items-top font-sans [&_a:hover]:underline col-span-12 text-s md:text-m font-heavy text-theme-text pb-xs"
        >
          <h2>
            headline!
          </h2>
        </div>
        <div
          class="body-copy col-span-12 text-body text-theme-text font-medium text-theme-body text-body"
        >
          <div>
            body copy of bv review submission
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
